File: backend/passservice.py
"""
业务编排层：参数解析 / 范围校验 / 过境计算 / 全局状态更新。

路由层（passesapi.py）只负责 HTTP 参数绑定与响应返回；
具体的业务流程——地面站预设解析、参数 clamp、TLE 获取、计算、
以及把结果写入 state._state 供 Socket.IO 实时广播——都收敛到这里，
避免路由层直接操纵其它模块的内部状态（原先散落在 passesapi._compute）。
"""

from __future__ import annotations

import threading
from dataclasses import asdict
from datetime import datetime, timezone

from config import (  # noqa: E402
    DEFAULT_ALT_M,
    DEFAULT_LAT,
    DEFAULT_LON,
    ON80DD_ALT_M,
    ON80DD_LAT,
    ON80DD_LON,
)
from astro import (  # noqa: E402
    PassesOutput,
    _time_to_dt,
    compute_groundtrack,
    compute_passes,
    passes_output_to_dict,
)
from state import _state  # noqa: E402
import tle  # noqa: E402
from tle import _get_tle_cached, _parse_tle_epoch, _resolve_satellite  # noqa: E402

# 全局实时状态写锁：请求改走线程池后，多个计算请求 / Socket.IO 广播可能并发
# 读写 state._state，用锁保证"站点配置 + 输出 + 当前卫星"作为整体原子更新，
# 避免广播读到一半更新（例如 station 是新的、output 还是旧的）。
_state_lock = threading.Lock()

MAX_HOURS = 24 * 14  # 最多 14 天
MIN_SAMPLE_SEC = 1
MAX_SAMPLE_SEC = 600
MIN_STEP_SEC = 10
MAX_STEP_SEC = 600


def _clamp(value: float, lo: float, hi: float) -> float:
    """把数值夹在 [lo, hi] 区间内，防止请求参数超出合理范围。"""
    return min(max(value, lo), hi)


def _resolve_station(params: dict) -> dict:
    """解析地面站坐标与内置预设（preset），返回 {lat, lon, alt, label}。

    坐标做范围 clamp，防止越界参数传给 skyfield 导致 500 或无效结果。
    """
    lat = _clamp(float(params.get("lat", DEFAULT_LAT)), -90.0, 90.0)
    lon = _clamp(float(params.get("lon", DEFAULT_LON)), -180.0, 180.0)
    alt = _clamp(float(params.get("alt", DEFAULT_ALT_M)), 0.0, 10000.0)
    preset = str(params.get("preset", "")).lower()
    if preset == "on80dd":
        lat, lon, alt = ON80DD_LAT, ON80DD_LON, ON80DD_ALT_M
        label = "ON80DD"
    elif preset in ("beijing", "bj"):
        lat, lon, alt = DEFAULT_LAT, DEFAULT_LON, DEFAULT_ALT_M
        label = "Beijing"
    else:
        label = f"{lat:.3f}, {lon:.3f}"
    return {"lat": lat, "lon": lon, "alt": alt, "label": label}


def _update_state(station: dict, sat_key: str, norad_id: int, hours: int,
                  sample_sec: int, horizon: float, output_dict: dict) -> None:
    """把最近一次计算写入 state._state，供 Socket.IO 实时位置广播只读使用。"""
    with _state_lock:
        _state["station"] = {
            "lat": station["lat"], "lon": station["lon"], "alt": station["alt"],
            "label": station["label"],
            "hours": hours, "sample_interval": sample_sec, "horizon": horizon,
            "satellite": sat_key, "norad_id": norad_id,
        }
        _state["output"] = output_dict
        _state["current_satellite"] = sat_key


def compute_passes_service(params: dict) -> dict:
    """过境业务编排：解析参数 → clamp 校验 → 计算 → 更新 state → 返回 dict。"""
    station = _resolve_station(params)
    hours = int(_clamp(float(params.get("hours", 48)), 1, MAX_HOURS))
    sample_sec = int(_clamp(float(params.get("sample_interval", 30)),
                            MIN_SAMPLE_SEC, MAX_SAMPLE_SEC))
    horizon = _clamp(float(params.get("horizon", 0.0)), -90.0, 90.0)
    sat_key, norad_id = _resolve_satellite(params)

    tle_name, tle1, tle2 = _get_tle_cached(norad_id)
    passes, sat = compute_passes(
        tle_name=tle_name, tle1=tle1, tle2=tle2,
        lat=station["lat"], lon=station["lon"], alt_m=station["alt"],
        hours=hours, horizon_deg=horizon, sample_interval_sec=sample_sec,
    )
    epoch_dt = _time_to_dt(sat.epoch) if sat.epoch is not None else None
    output = PassesOutput(
        satellite_name=tle_name,
        norad_id=norad_id,
        tle_name=tle_name,
        tle1=tle1,
        tle2=tle2,
        tle_epoch=epoch_dt.isoformat() if epoch_dt else "",
        station_lat=round(station["lat"], 6),
        station_lon=round(station["lon"], 6),
        station_alt_m=round(station["alt"], 2),
        station_label=station["label"],
        generated_at=datetime.now(timezone.utc).isoformat(),
        horizon_deg=horizon,
        hours=hours,
        sample_interval_sec=sample_sec,
        passes=passes,
    )
    output_dict = passes_output_to_dict(output)
    # 附加 TLE 来源标记（online / fallback / cache），供前端透明提示数据新鲜度
    output_dict["tle_source"] = tle.tle_source(norad_id)

    _update_state(station, sat_key, norad_id, hours, sample_sec, horizon, output_dict)
    return output_dict


def compute_groundtrack_service(params: dict) -> dict:
    """星下点轨迹业务编排：解析参数 → clamp 校验 → 计算 → 返回 dict。"""
    station = _resolve_station(params)
    hours = int(_clamp(float(params.get("hours", 48)), 1, MAX_HOURS))
    step_sec = int(_clamp(float(params.get("step_sec", 60)),
                          MIN_STEP_SEC, MAX_STEP_SEC))

    sat_key, norad_id = _resolve_satellite(params)
    tle_name, tle1, tle2 = _get_tle_cached(norad_id)
    points = compute_groundtrack(
        tle_name=tle_name, tle1=tle1, tle2=tle2,
        lat=station["lat"], lon=station["lon"], alt_m=station["alt"],
        hours=hours, step_sec=step_sec,
    )
    epoch_dt = _parse_tle_epoch(tle1)
    return {
        "satellite_name": tle_name,
        "norad_id": norad_id,
        "tle_epoch": epoch_dt.isoformat() if epoch_dt else "",
        "tle_source": tle.tle_source(norad_id),
        "station_label": station["label"],
        "station_lat": round(station["lat"], 6),
        "station_lon": round(station["lon"], 6),
        "station_alt_m": round(station["alt"], 2),
        "generated_at": datetime.now(timezone.utc).isoformat(),
        "hours": hours,
        "step_sec": step_sec,
        "points": [asdict(p) for p in points],
    }

File: backend-rs/src/passservice.rs
//! 业务编排层：对应 backend/passservice.py。
//!
//! 参数解析 / 范围校验 / 过境计算 / 全局状态更新。路由层只负责 HTTP 参数绑定；
//! 具体业务流程（站点预设解析、clamp、TLE 获取、计算、写 state 供实时广播）收敛到这里。
//! 计算为 CPU 密集，放到 `spawn_blocking` 执行（对齐 Python 由 FastAPI 线程池运行同步 handler）。

use chrono::Utc;
use serde_json::{Value, json};

use crate::astroconv::iso_utc;
use crate::config;
use crate::error::ApiError;
use crate::models::{GroundTrackResponse, PassesResponse};
use crate::{astro, state, tle};

pub const MAX_HOURS: i64 = 24 * 14; // 最多 14 天
pub const MIN_SAMPLE_SEC: i64 = 1;
pub const MAX_SAMPLE_SEC: i64 = 600;
pub const MIN_STEP_SEC: i64 = 10;
pub const MAX_STEP_SEC: i64 = 600;

fn clamp(value: f64, lo: f64, hi: f64) -> f64 {
    value.max(lo).min(hi)
}

fn num(params: &Value, key: &str, default: f64) -> f64 {
    params.get(key).and_then(|v| v.as_f64()).unwrap_or(default)
}

/// 解析地面站坐标与内置预设（preset），返回 (lat, lon, alt, label)。
fn resolve_station(params: &Value) -> (f64, f64, f64, String) {
    let mut lat = clamp(num(params, "lat", *config::DEFAULT_LAT), -90.0, 90.0);
    let mut lon = clamp(num(params, "lon", *config::DEFAULT_LON), -180.0, 180.0);
    let mut alt = clamp(num(params, "alt", *config::DEFAULT_ALT_M), 0.0, 10000.0);
    let preset = params
        .get("preset")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_lowercase();
    let label;
    if preset == "on80dd" {
        lat = *config::ON80DD_LAT;
        lon = *config::ON80DD_LON;
        alt = *config::ON80DD_ALT_M;
        label = "ON80DD".to_string();
    } else if preset == "beijing" || preset == "bj" {
        lat = *config::DEFAULT_LAT;
        lon = *config::DEFAULT_LON;
        alt = *config::DEFAULT_ALT_M;
        label = "Beijing".to_string();
    } else {
        label = format!("{:.3}, {:.3}", lat, lon);
    }
    (lat, lon, alt, label)
}

/// 把计算/TLE 错误映射为通用 500（对齐 Python 未处理异常的兜底行为），并记录详情。
fn internal_err(context: &str, err: String) -> ApiError {
    tracing::error!("{}: {}", context, err);
    ApiError::service("服务器内部错误")
}

/// 过境业务编排：解析参数 → clamp 校验 → 计算 → 更新 state → 返回响应。
pub async fn compute_passes_service(params: Value) -> Result<PassesResponse, ApiError> {
    let (lat, lon, alt, label) = resolve_station(&params);
    let hours = clamp(num(&params, "hours", 48.0), 1.0, MAX_HOURS as f64) as i64;
    let sample_sec = clamp(
        num(&params, "sample_interval", 30.0),
        MIN_SAMPLE_SEC as f64,
        MAX_SAMPLE_SEC as f64,
    ) as i64;
    let horizon = clamp(num(&params, "horizon", 0.0), -90.0, 90.0);
    let (sat_key, norad_id) = tle::resolve_satellite(&params);

    let (tle_name, tle1, tle2) = tle::get_tle_cached(norad_id)
        .await
        .map_err(|e| internal_err("获取 TLE 失败", e))?;

    // CPU 密集计算放到阻塞线程池，避免卡住事件循环
    let c_name = tle_name.clone();
    let c_l1 = tle1.clone();
    let c_l2 = tle2.clone();
    let passes = tokio::task::spawn_blocking(move || {
        astro::compute_passes(&c_name, &c_l1, &c_l2, lat, lon, alt, hours, horizon, sample_sec)
    })
    .await
    .map_err(|e| internal_err("计算任务失败", e.to_string()))?
    .map_err(|e| internal_err("过境计算失败", e))?;

    let epoch = tle::parse_tle_epoch(&tle1)
        .map(|d| iso_utc(&d))
        .unwrap_or_default();
    let source = tle::tle_source(norad_id);

    let output = PassesResponse {
        satellite_name: tle_name.clone(),
        norad_id,
        tle_name: tle_name.clone(),
        tle1: tle1.clone(),
        tle2: tle2.clone(),
        tle_epoch: epoch,
        station_lat: astro::round(lat, 6),
        station_lon: astro::round(lon, 6),
        station_alt_m: astro::round(alt, 2),
        station_label: label.clone(),
        generated_at: iso_utc(&Utc::now()),
        horizon_deg: horizon,
        hours,
        sample_interval_sec: sample_sec,
        tle_source: source,
        passes,
    };

    // 写入运行时状态，供 Socket.IO 实时位置广播只读使用
    let station_state = json!({
        "lat": lat, "lon": lon, "alt": alt, "label": label,
        "hours": hours, "sample_interval": sample_sec, "horizon": horizon,
        "satellite": sat_key.clone(), "norad_id": norad_id,
    });
    let output_value = serde_json::to_value(&output).unwrap_or_else(|_| json!({}));
    state::update_state(station_state, output_value, sat_key);

    Ok(output)
}

/// 星下点轨迹业务编排：解析参数 → clamp 校验 → 计算 → 返回响应（不更新 state）。
pub async fn compute_groundtrack_service(params: Value) -> Result<GroundTrackResponse, ApiError> {
    let (lat, lon, alt, label) = resolve_station(&params);
    let hours = clamp(num(&params, "hours", 48.0), 1.0, MAX_HOURS as f64) as i64;
    let step_sec = clamp(
        num(&params, "step_sec", 60.0),
        MIN_STEP_SEC as f64,
        MAX_STEP_SEC as f64,
    ) as i64;

    let (_sat_key, norad_id) = tle::resolve_satellite(&params);
    let (tle_name, tle1, tle2) = tle::get_tle_cached(norad_id)
        .await
        .map_err(|e| internal_err("获取 TLE 失败", e))?;

    let c_name = tle_name.clone();
    let c_l1 = tle1.clone();
    let c_l2 = tle2.clone();
    let points = tokio::task::spawn_blocking(move || {
        astro::compute_groundtrack(&c_name, &c_l1, &c_l2, lat, lon, alt, hours, step_sec)
    })
    .await
    .map_err(|e| internal_err("计算任务失败", e.to_string()))?
    .map_err(|e| internal_err("轨迹计算失败", e))?;

    let epoch = tle::parse_tle_epoch(&tle1)
        .map(|d| iso_utc(&d))
        .unwrap_or_default();

    Ok(GroundTrackResponse {
        satellite_name: tle_name,
        norad_id,
        tle_epoch: epoch,
        tle_source: tle::tle_source(norad_id),
        station_label: label,
        station_lat: astro::round(lat, 6),
        station_lon: astro::round(lon, 6),
        station_alt_m: astro::round(alt, 2),
        generated_at: iso_utc(&Utc::now()),
        hours,
        step_sec,
        points,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn station_preset_on80dd() {
        let (lat, lon, _alt, label) = resolve_station(&json!({ "preset": "on80dd" }));
        assert_eq!(label, "ON80DD");
        assert!((lat - 40.1458).abs() < 1e-6);
        assert!((lon - 116.2917).abs() < 1e-6);
    }

    #[test]
    fn station_preset_beijing_alias() {
        let (_lat, _lon, _alt, label) = resolve_station(&json!({ "preset": "BJ" }));
        assert_eq!(label, "Beijing");
    }

    #[test]
    fn station_clamp_and_label() {
        let (lat, lon, _alt, label) = resolve_station(&json!({ "lat": 999.0, "lon": -999.0 }));
        assert_eq!(lat, 90.0);
        assert_eq!(lon, -180.0);
        assert_eq!(label, "90.000, -180.000");
    }
}
